File: ebitengine/ui/hud.go
package ui

import (
	"fmt"
	"image/color"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/text/v2"
	"github.com/hajimehoshi/ebiten/v2/vector"
	"github.com/thesimpledev/radianthorizon/core"
	"github.com/thesimpledev/radianthorizon/entities"
	"github.com/thesimpledev/radianthorizon/game"
)

const (
	infoBarH = 60
	infoBarY = core.WindowH - infoBarH
)

// HUD draws the heads-up display: HP bar, deck count, weapon info, skip indicator, card info.
type HUD struct {
	Font      *text.GoTextFace
	SmallFont *text.GoTextFace
}

// NewHUD creates the HUD with fonts.
func NewHUD(font, smallFont *text.GoTextFace) *HUD {
	return &HUD{Font: font, SmallFont: smallFont}
}

// Draw renders the full HUD.
func (h *HUD) Draw(screen *ebiten.Image, hp, maxHP, deckRemaining int, ws *entities.WeaponSlot, canAvoid bool, phase string, hoveredCard *entities.Card, room *game.Room) {
	h.drawHPBar(screen, hp, maxHP)
	h.drawDeckCount(screen, deckRemaining)
	h.drawWeaponInfo(screen, ws)
	h.drawSkipIndicator(screen, canAvoid)
	h.drawCardInfo(screen, hoveredCard, ws, room, hp)
}

func (h *HUD) drawHPBar(screen *ebiten.Image, hp, maxHP int) {
	x, y := float32(20), float32(20)
	w, h2 := float32(200), float32(24)
	fill := (float32(hp) / float32(maxHP)) * w

	// Background
	vector.DrawFilledRect(screen, x, y, w, h2, core.ColorHPBG, true)

	// Fill
	var fillColor color.NRGBA
	if hp <= 5 {
		fillColor = core.ColorRed
	} else {
		fillColor = core.ColorHPBar
	}
	vector.DrawFilledRect(screen, x, y, fill, h2, fillColor, true)

	// Border
	vector.StrokeRect(screen, x, y, w, h2, 1, core.ColorWhite, true)

	// Text
	op := &text.DrawOptions{}
	op.GeoM.Translate(float64(x)+8, float64(y)+4)
	op.ColorScale.ScaleWithColor(core.ColorWhite)
	text.Draw(screen, fmt.Sprintf("HP: %d / %d", hp, maxHP), h.Font, op)
}

func (h *HUD) drawDeckCount(screen *ebiten.Image, remaining int) {
	op := &text.DrawOptions{}
	op.GeoM.Translate(core.DungeonX, core.DungeonY-22)
	op.ColorScale.ScaleWithColor(core.ColorWhite)
	text.Draw(screen, fmt.Sprintf("Dungeon: %d", remaining), h.Font, op)
}

func (h *HUD) drawWeaponInfo(screen *ebiten.Image, ws *entities.WeaponSlot) {
	x := float64(core.WeaponX)
	y := float64(core.WeaponY) - 22

	op := &text.DrawOptions{}
	op.GeoM.Translate(x, y)
	op.ColorScale.ScaleWithColor(core.ColorWhite)

	if ws.HasWeapon() {
		weapon := ws.Weapon
		text.Draw(screen, fmt.Sprintf("Weapon: %s of %s", weapon.Rank, weapon.Suit), h.Font, op)

		if len(ws.Slain) > 0 {
			last := ws.Slain[len(ws.Slain)-1]
			op2 := &text.DrawOptions{}
			slainY := y + core.CardDrawH + float64(len(ws.Slain))*core.SlainOffsetY + 10
			op2.GeoM.Translate(x, slainY)
			op2.ColorScale.ScaleWithColor(core.ColorWhite)
			text.Draw(screen, fmt.Sprintf("Last slain: %s (can fight \u2264%d)", last.Rank, last.Value), h.Font, op2)
		}
	} else {
		text.Draw(screen, "No weapon equipped", h.Font, op)
	}
}

func (h *HUD) drawSkipIndicator(screen *ebiten.Image, canAvoid bool) {
	x := float64(core.WindowW - 220)
	y := float64(20)

	op := &text.DrawOptions{}
	op.GeoM.Translate(x, y)

	if canAvoid {
		op.ColorScale.ScaleWithColor(core.ColorGold)
		text.Draw(screen, "[ Press S to skip room ]", h.Font, op)
	} else {
		op.ColorScale.ScaleWithColor(color.NRGBA{153, 153, 153, 153})
		text.Draw(screen, "Cannot skip this room", h.Font, op)
	}
}

func (h *HUD) drawCardInfo(screen *ebiten.Image, card *entities.Card, ws *entities.WeaponSlot, room *game.Room, hp int) {
	// Background bar
	vector.DrawFilledRect(screen, 0, float32(infoBarY), core.WindowW, infoBarH, color.NRGBA{0, 0, 0, 153}, true)
	vector.StrokeRect(screen, 0, float32(infoBarY), core.WindowW, infoBarH, 1, color.NRGBA{102, 102, 102, 128}, true)

	if card == nil {
		op := &text.DrawOptions{}
		op.GeoM.Translate(20, float64(infoBarY)+8)
		op.ColorScale.ScaleWithColor(color.NRGBA{128, 128, 128, 128})
		text.Draw(screen, "Hover over a card to see what it does", h.Font, op)
		return
	}

	name := card.Rank + " of " + card.Suit
	var line1, line2 string

	if card.IsMonster {
		line1, line2 = h.monsterInfo(card, ws)
	} else if card.IsWeapon {
		line1, line2 = h.weaponInfoText(card, ws)
	} else if card.IsPotion {
		line1, line2 = h.potionInfo(card, room, hp)
	}

	// Card name with type color
	nameOp := &text.DrawOptions{}
	nameOp.GeoM.Translate(20, float64(infoBarY)+8)
	if card.IsMonster {
		nameOp.ColorScale.ScaleWithColor(core.ColorRed)
	} else if card.IsWeapon {
		nameOp.ColorScale.ScaleWithColor(color.NRGBA{153, 179, 255, 255})
	} else if card.IsPotion {
		nameOp.ColorScale.ScaleWithColor(core.ColorHPBar)
	}
	text.Draw(screen, name, h.Font, nameOp)

	// Description line 1
	nameW, _ := text.Measure(name, h.Font, 0)
	if line1 != "" {
		op := &text.DrawOptions{}
		op.GeoM.Translate(20+nameW+20, float64(infoBarY)+8)
		op.ColorScale.ScaleWithColor(core.ColorWhite)
		text.Draw(screen, line1, h.Font, op)
	}

	// Description line 2
	if line2 != "" {
		op := &text.DrawOptions{}
		op.GeoM.Translate(20, float64(infoBarY)+30)
		op.ColorScale.ScaleWithColor(color.NRGBA{204, 204, 204, 204})
		text.Draw(screen, line2, h.Font, op)
	}
}

func (h *HUD) monsterInfo(card *entities.Card, ws *entities.WeaponSlot) (string, string) {
	line1 := fmt.Sprintf("Monster  |  Deals %d damage", card.Value)
	parts := fmt.Sprintf("Barehanded: -%d HP", card.Value)

	if ws.HasWeapon() {
		if ws.CanUseAgainst(card.Value) {
			weapDmg := card.Value - ws.WeaponValue()
			if weapDmg < 0 {
				weapDmg = 0
			}
			if weapDmg == 0 {
				parts += "    Weapon: no damage!"
			} else {
				parts += fmt.Sprintf("    Weapon: -%d HP", weapDmg)
			}
		} else {
			last := ws.Slain[len(ws.Slain)-1]
			parts += fmt.Sprintf("    Weapon: blocked (monster %d > last slain %d)", card.Value, last.Value)
		}
	}

	return line1, parts
}

func (h *HUD) weaponInfoText(card *entities.Card, ws *entities.WeaponSlot) (string, string) {
	line1 := fmt.Sprintf("Weapon  |  Absorbs up to %d damage per fight", card.Value)
	var line2 string

	if ws.HasWeapon() {
		old := ws.Weapon
		if card.Value > old.Value {
			line2 = fmt.Sprintf("Replaces current weapon (%s of %s) - upgrade!", old.Rank, old.Suit)
		} else if card.Value < old.Value {
			line2 = fmt.Sprintf("Replaces current weapon (%s of %s) - downgrade", old.Rank, old.Suit)
		} else {
			line2 = "Replaces current weapon (same strength, resets slain stack)"
		}
	} else {
		line2 = "You have no weapon - equipping this lets you fight monsters with reduced damage"
	}

	return line1, line2
}

func (h *HUD) potionInfo(card *entities.Card, room *game.Room, hp int) (string, string) {
	line1 := fmt.Sprintf("Health Potion  |  Heals %d HP (max %d)", card.Value, core.MaxHP)
	var line2 string

	if room.PotionUsed {
		line2 = "Already used a potion this room - this one will be wasted!"
	} else {
		actualHeal := card.Value
		if core.MaxHP-hp < actualHeal {
			actualHeal = core.MaxHP - hp
		}
		if actualHeal <= 0 {
			line2 = "HP is already full - healing would be wasted"
		} else if actualHeal < card.Value {
			line2 = fmt.Sprintf("Would heal %d HP (capped at max %d)", actualHeal, core.MaxHP)
		} else {
			line2 = fmt.Sprintf("Would heal %d HP  (%d -> %d)", actualHeal, hp, hp+actualHeal)
		}
	}

	return line1, line2
}

File: love/ui/hud.lua
-- ui/hud.lua
-- Heads-up display: HP bar, deck count, weapon info, skip indicator,
-- and a card info bar at the bottom of the screen.

local C     = require("core.constants")
local Rules = require("game.rules")

local HUD = {}

-- Info bar dimensions
local INFO_BAR_H = 60
local INFO_BAR_Y = C.WINDOW_H - INFO_BAR_H

-- Draw the full HUD. Called every frame from game_play:draw().
function HUD.draw(hp, max_hp, deck_remaining, weapon_slot, can_avoid, phase, hovered_card, room)
    HUD.draw_hp_bar(hp, max_hp)
    HUD.draw_deck_count(deck_remaining)
    HUD.draw_weapon_info(weapon_slot)
    HUD.draw_skip_indicator(can_avoid)
    HUD.draw_card_info(hovered_card, weapon_slot, room, hp)
end

-- HP bar in the top-left corner
function HUD.draw_hp_bar(hp, max_hp)
    local x, y = 20, 20
    local w, h = 200, 24
    local fill = (hp / max_hp) * w

    -- Background
    love.graphics.setColor(C.COLOR_HP_BG)
    love.graphics.rectangle("fill", x, y, w, h, 4, 4)

    -- Fill (color shifts red when low)
    if hp <= 5 then
        love.graphics.setColor(C.COLOR_RED)
    else
        love.graphics.setColor(C.COLOR_HP_BAR)
    end
    love.graphics.rectangle("fill", x, y, fill, h, 4, 4)

    -- Border
    love.graphics.setColor(C.COLOR_WHITE)
    love.graphics.rectangle("line", x, y, w, h, 4, 4)

    -- Text
    love.graphics.setColor(C.COLOR_WHITE)
    love.graphics.print("HP: " .. hp .. " / " .. max_hp, x + 8, y + 4)
end

-- Deck count near the dungeon pile
function HUD.draw_deck_count(remaining)
    love.graphics.setColor(C.COLOR_WHITE)
    love.graphics.print("Dungeon: " .. remaining, C.DUNGEON_X, C.DUNGEON_Y - 22)
end

-- Weapon info below the weapon slot
function HUD.draw_weapon_info(weapon_slot)
    local x = C.WEAPON_X
    local y = C.WEAPON_Y - 22

    love.graphics.setColor(C.COLOR_WHITE)
    if weapon_slot:has_weapon() then
        local weapon = weapon_slot.weapon
        love.graphics.print("Weapon: " .. weapon.rank .. " of " .. weapon.suit, x, y)

        if #weapon_slot.slain > 0 then
            local last = weapon_slot.slain[#weapon_slot.slain]
            love.graphics.print(
                "Last slain: " .. last.rank .. " (can fight \u{2264}" .. last.value .. ")",
                x, y + C.CARD_DRAW_H + #weapon_slot.slain * C.SLAIN_OFFSET_Y + 10
            )
        end
    else
        love.graphics.print("No weapon equipped", x, y)
    end
end

-- "Skip available" / "Must face this room" indicator
function HUD.draw_skip_indicator(can_avoid)
    local x = C.WINDOW_W - 220
    local y = 20

    if can_avoid then
        love.graphics.setColor(C.COLOR_GOLD)
        love.graphics.print("[ Press S to skip room ]", x, y)
    else
        love.graphics.setColor(0.6, 0.6, 0.6, 0.6)
        love.graphics.print("Cannot skip this room", x, y)
    end
end

-- Card info bar at the bottom of the screen.
-- Shows what the hovered card does in the current game context.
function HUD.draw_card_info(card, weapon_slot, room, hp)
    -- Background bar (always visible so layout is stable)
    love.graphics.setColor(0, 0, 0, 0.6)
    love.graphics.rectangle("fill", 0, INFO_BAR_Y, C.WINDOW_W, INFO_BAR_H)
    love.graphics.setColor(0.4, 0.4, 0.4, 0.5)
    love.graphics.rectangle("line", 0, INFO_BAR_Y, C.WINDOW_W, INFO_BAR_H)

    if not card then
        love.graphics.setColor(0.5, 0.5, 0.5, 0.5)
        love.graphics.print("Hover over a card to see what it does", 20, INFO_BAR_Y + 8)
        return
    end

    local name = card.rank .. " of " .. card.suit
    local line1, line2

    if card.is_monster then
        line1, line2 = HUD.monster_info(card, weapon_slot, room)
    elseif card.is_weapon then
        line1, line2 = HUD.weapon_info_text(card, weapon_slot)
    elseif card.is_potion then
        line1, line2 = HUD.potion_info(card, room, hp)
    end

    -- Card name (highlighted by type color)
    if card.is_monster then
        love.graphics.setColor(C.COLOR_RED)
    elseif card.is_weapon then
        love.graphics.setColor(0.6, 0.7, 1.0, 1)
    elseif card.is_potion then
        love.graphics.setColor(C.COLOR_HP_BAR)
    end
    love.graphics.print(name, 20, INFO_BAR_Y + 8)

    -- Description lines
    local name_w = love.graphics.getFont():getWidth(name)
    love.graphics.setColor(C.COLOR_WHITE)
    if line1 then
        love.graphics.print(line1, 20 + name_w + 20, INFO_BAR_Y + 8)
    end
    if line2 then
        love.graphics.setColor(0.8, 0.8, 0.8, 0.8)
        love.graphics.print(line2, 20, INFO_BAR_Y + 30)
    end
end

-- Build info text for a monster card.
function HUD.monster_info(card, weapon_slot, room)
    local line1 = "Monster  |  Deals " .. card.value .. " damage"
    local parts = {}

    -- Barehanded option (always available)
    table.insert(parts, "Barehanded: -" .. card.value .. " HP")

    -- Weapon option
    if weapon_slot:has_weapon() then
        if weapon_slot:can_use_against(card.value) then
            local weap_dmg = math.max(0, card.value - weapon_slot:weapon_value())
            if weap_dmg == 0 then
                table.insert(parts, "Weapon: no damage!")
            else
                table.insert(parts, "Weapon: -" .. weap_dmg .. " HP")
            end
        else
            local last = weapon_slot.slain[#weapon_slot.slain]
            table.insert(parts, "Weapon: blocked (monster " .. card.value .. " > last slain " .. last.value .. ")")
        end
    end

    return line1, table.concat(parts, "    ")
end

-- Build info text for a weapon card.
function HUD.weapon_info_text(card, weapon_slot)
    local line1 = "Weapon  |  Absorbs up to " .. card.value .. " damage per fight"
    local line2

    if weapon_slot:has_weapon() then
        local old = weapon_slot.weapon
        if card.value > old.value then
            line2 = "Replaces current weapon (" .. old.rank .. " of " .. old.suit .. ") - upgrade!"
        elseif card.value < old.value then
            line2 = "Replaces current weapon (" .. old.rank .. " of " .. old.suit .. ") - downgrade"
        else
            line2 = "Replaces current weapon (same strength, resets slain stack)"
        end
    else
        line2 = "You have no weapon - equipping this lets you fight monsters with reduced damage"
    end

    return line1, line2
end

-- Build info text for a potion card.
function HUD.potion_info(card, room, hp)
    local line1 = "Health Potion  |  Heals " .. card.value .. " HP (max " .. C.MAX_HP .. ")"
    local line2

    if room.potion_used then
        line2 = "Already used a potion this room - this one will be wasted!"
    else
        local actual_heal = math.min(card.value, C.MAX_HP - hp)
        if actual_heal <= 0 then
            line2 = "HP is already full - healing would be wasted"
        elseif actual_heal < card.value then
            line2 = "Would heal " .. actual_heal .. " HP (capped at max " .. C.MAX_HP .. ")"
        else
            line2 = "Would heal " .. actual_heal .. " HP  (" .. hp .. " -> " .. (hp + actual_heal) .. ")"
        end
    end

    return line1, line2
end

return HUD
